test/test_torch_implementations.py

```diff
@@ -0,0 +1,379 @@
+        )
+
+
+def test_full_network_architectures():
+    """Test that full Class1NeuralNetwork and TorchNeuralNetwork implementations match."""
+    
+    # Test different architectures
+    architectures = [
+        {
+            # Basic architecture
+            "peptide_encoding": {
+                "vector_encoding_name": "BLOSUM62",
+                "alignment_method": "pad_middle",
+                "max_length": 15,
+            },
+            "layer_sizes": [64, 32],
+            "dropout_probability": 0.0,
+            "batch_normalization": True,
+            "locally_connected_layers": [],
+            "activation": "tanh",
+            "init": "glorot_uniform",
+            "output_activation": "sigmoid",
+        },
+    ]
+
+    for arch_params in architectures:
+        # Create Keras model
+        keras_model = Class1NeuralNetwork(**arch_params)
+        
+        # Create test input to force network initialization
+        test_peptides = ["SIINFEKL", "KLGGALQAK"]
+        peptide_encoding = EncodableSequences.create(test_peptides)
+        
+        # Initialize network explicitly
+        keras_model._network = keras_model.make_network(
+            **keras_model.network_hyperparameter_defaults.subselect(keras_model.hyperparameters)
+        )
+        
+        # Now we can safely get and compile the network
+        network = keras_model.network()
+        assert network is not None, "Network initialization failed"
+        network.compile(optimizer='adam', loss='mse')
+
+        # Create equivalent PyTorch model
+        torch_model = TorchNeuralNetwork(**arch_params)
+
+        # Debugging: Print Keras and Torch layer shapes
+        print("\n[DEBUG] Keras model layers:")
+        for idx, k_layer in enumerate(network.layers):
+            k_weights = k_layer.get_weights()
+            shapes = [w.shape for w in k_weights]
+            print(f"  Keras layer #{idx}: {k_layer.__class__.__name__} weight shapes={shapes}")
+
+        print("\n[DEBUG] Torch model (dense_layers + output_layer):")
+        combined_layers = list(torch_model.dense_layers) + [torch_model.output_layer]
+        for idx, t_layer in enumerate(combined_layers):
+            if hasattr(t_layer, 'weight') and t_layer.weight is not None:
+                print(f"  Torch layer #{idx}: {t_layer.__class__.__name__} weight shape={tuple(t_layer.weight.shape)}, bias shape={tuple(t_layer.bias.shape)}")
+            else:
+                print(f"  Torch layer #{idx}: {t_layer.__class__.__name__} (no linear weights)")
+        torch_model.load_weights_from_keras(network)
+
+        # Get predictions from both models
+        keras_predictions = keras_model.predict(peptides=peptide_encoding)
+        torch_predictions = torch_model.predict(peptides=peptide_encoding)
+
+        # Compare predictions
+        assert_array_almost_equal(
+            keras_predictions,
+            torch_predictions,
+            decimal=0,  # or add atol=1.0
```

Comment:
@timodonnell This is a big tolerance because the outputs are in the hundreds - thousands range... Let me know if this is a problem. I couldn't get it down to smaller tolerance.

Reply:
I don't love it but it's on a log scale so for very high values 1 nM difference may correspond to a pretty small difference in the underlying neural network output. I think it's not necessarily a big problem but I'd ideally like to understand better where the discrepancies are arising in the network

Reply:
It's this: `max_ic50 ** (1.0 - x)`.  The regression targets themselves are within 4 decimal places but once they go through that, things get wonky. Not sure what to do here. Any ideas?